### ISSUE TITLE
NPP crashes on hiding lines

### ISSUE THREAD

**Issue by Yaron10** [notepad-plus-plus/notepad-plus-plus#12184]
**STR:**
Paste the following lines in a new document.
```
Test
Test
Test
Test
Test
```
Place the caret at line 3.
View -> Hide Lines.
Place the caret at line 2.
View -> Hide Lines.
Click on the show-lines markers.
You should get this:
![תמונה](https://user-images.githubusercontent.com/10229320/190706586-710c83b2-2ff7-4a34-ae4b-4b777c20d280.png)
Click on the single show-lines marker.

**Result:**
NPP crashes.

***

Notepad++ v8.4.5   (64-bit)
Build time : Sep 16 2022 - 07:53:18
Path : C:\***Notepad++ Original.exe
Command Line : 
Admin mode : ON
Local Conf mode : ON
Cloud Config : OFF
OS Name : Windows 10 Pro (64-bit) 
OS Version : 20H2
OS Build : 19042.985
Current ANSI codepage : 1255
Plugins : 
    ComparePlus (1)
    HTMLTag (1.3.2)
    MenuIcons (1.2.5)
    PythonScript (2)
    _CustomizeToolbar (4.2)

**Comment by alankilborn** [comment]
I can replicate.

**Comment by xomx** [comment]
(Almost) infinite [loop](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/3295cc8d2c8eccc303861a19fbfd0e5910f24ddf/PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp#L3490) caused by unsigned integer overflow.

This specific one can be fixed by changing the `size_t lineNumber` param to `intptr_t lineNumber`.

Otherwise in the

`for (lineNumber--; lineNumber >= 0 && !openPresent; lineNumber--)`

when the `lineNumber` reaches 0, the next `lineNumber--` operation causes unsigned integer overflow, so the `lineNumber` becomes 0xffffffffffffffff and the `lineNumber >= 0` condition is then true (`lineNumber` is not now the intended -1 but 18446744073709551615 in decimal ...) and the for-cycle does not end at this point as it should.

**Comment by chcg** [comment]
Probably also doable with `size_t `and `lineNumber > 0`, but `linenumber-1` in the loop itself.

**Comment by Yaron10** [comment]
Thank you @xomx and @chcg. 👍 

@xomx,

> This specific one can be fixed by changing the size_t lineNumber param to intptr_t lineNumber.

You suspect it might not be the only one, correct?
Are you planning to review some code before submitting a PR for this issue?

**Comment by xomx** [comment]
@Yaron10 

You can really read between the lines :-)

Moreover  when I fixed this integer overflow, the green symbol on your pic did not show back the line No.3, it has simply stayed hidden (exactly in the state as is shown on your pic here) until N++ restart.

As I am not using this feature at all, I am slightly reluctant to spend my time on that. Is any Hide Lines feature backer here willing to do that review?

**Comment by Yaron10** [comment]
@xomx,

> You can really read between the lines :-)

If the lines are _not hidden_. :)

> Moreover when I fixed this integer overflow, the green symbol on your pic did not show back the line No.3, it has simply stayed hidden (exactly in the state as is shown on your pic here) until N++ restart.

👍 

> As I am not using this feature at all, I am slightly reluctant to spend my time on that.

I understand and respect that.
I don't use this feature either.

Thanks again.

**Comment by alankilborn** [comment]
@xomx 

>Moreover when I fixed this integer overflow, the green symbol on your pic did not show back the line No.3, it has simply stayed hidden (exactly in the state as is shown on your pic here) until N++ restart.

The hide lines feature currently implemented has a lot of such problems.
I am attempting to "guide development" :-) to make it a workable feature.

>As I am not using this feature at all, I am slightly reluctant to spend my time on that.

It would be appreciated if you'd fix (officially) only the crash part, as (a) software should never crash, and (b) it sounds like you've already found the fix.

And...if hidden-lines was a workable feature, and had an interface to _Find All_, you just might find some uses for it. :-)

**Comment by Adrian104** [comment]
> (Almost) infinite [loop](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/3295cc8d2c8eccc303861a19fbfd0e5910f24ddf/PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp#L3490) caused by unsigned integer overflow.

I can also confirm that overflow really occurs there, crashing the whole program.
However, I think it is not the cause, but rather a side effect of `hideLines()` not working properly in specific circumstances. In other words, hiding lines can introduce invalid state of green markers. Clicking on them may lead to unexpected behavior (including crash) because of that invalid state. By invalid state, I mean this:

![img](https://user-images.githubusercontent.com/63365428/191832243-a2133ebd-b393-40d9-9886-14d2acc86b13.png)

A - this is what user sees
B - this is what program does under the hood
C - after clicking on first marker, markers 1 and 3 are removed and this is the result

I will try to fix this bug and (if everything works) I will make a PR.

**Comment by alankilborn** [comment]
@Adrian104 

>hiding lines can introduce invalid state of green markers

This is long known to be the case; see e.g. #8149

Personally, I am campaigning for those green margin symbols to be removed from the software, and only the green line (stretching across the editing area) remaining to show hidden lines.  See #12178 and #12109.

**Comment by vinsworldcom** [comment]
> Personally, I am campaigning for those green margin symbols to be removed from the software, and only the green line (stretching across the editing area) remaining to show hidden lines. See #12178 and #12109.

Less clutter is good, but it would make unhiding ?difficult?.  How would you unhide?  I don't think clicking the line in the Scintilla edit view works.  Scintilla margins can be click sensitive; hence the markers.  But maybe removing the bottom left-pointing arrow and only leaving the top right-pointing arrow (in Scintilla terms, this is a "marker" and it would be on the same Scintilla line number as the green underline, which is also a "marker") would work?  This may also eliminate situations where you have overlapping hides and see the unmatched left/right pointing arrows as some screenshots above.

I should also state I **never** use "Hide Lines" and it gave me loads of trouble working on a line number plugin that I just abandoned it.  [It also interferes with the existing Custom Line Numbers plugin](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/12046#issuecomment-1234263963).

Cheers.

**Comment by vinsworldcom** [comment]
> a line could be unhidden by putting the caret on the line and invoking a (new) show-lines command, e.g. shortcutted for speed

Good Point!  Currently, there is no unhide and the only way to do it is click the margin.  I do like mouse-less operation options!
